src/app/zeko/circuits/indexed_merkle_tree.ml

```diff
@@ -62,23 +79,25 @@ struct
       implied_root_raw Field.(constant typ zero) path_y
     in
     let* root_intermediate =
-      match check with
-      | Some check ->
-          if_ check ~typ:F.typ ~then_:root_intermediate
-            ~else_:root_intermediate'
-      | None ->
-          Checked.return root_intermediate
+      if_ check ~typ:F.typ ~then_:root_intermediate ~else_:root_intermediate'
     in
     let* () =
       assert_equal ~label:__LOC__ F.typ root_intermediate root_intermediate'
     in
     let* root_new = implied_root { key = y; next_key = z } path_y in
-    let* root =
-      match check with
-      | Some check ->
-          if_ check ~typ:F.typ ~then_:root ~else_:root_new
-      | None ->
-          Checked.return root
+    let* root = if_ check ~typ:F.typ ~then_:root ~else_:root_new in
+    (* Check that no empty indices have been skipped. *)
+    let* is_y_most_left =
+      foldl (List.zip_exn path_y empty_path) ~init:Boolean.true_
+        ~f:(fun acc (PathStep.{ hash_other; is_right }, empty_hash) ->
+          let* is_valid_left =
+            Field.Checked.equal empty_hash hash_other >>= Boolean.( &&& ) acc
+          in
+          if_ is_right ~typ:Boolean.typ ~then_:acc ~else_:is_valid_left )
```

Comment:
if is_right is true for all PathStep, then acc is always returned, and it's true, right? Isn't this wrong? Am I missing something?

Reply:
Yeah good catch. I need `is_valid_right` which is supposed to be not equal to the empty hash.